core/src/transport/global_only.rs

```diff
@@ -0,0 +1,353 @@
+// Copyright 2023 Protocol Labs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError, TransportEvent},
+    Transport,
+};
+use log::debug;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Dropping all dial requests to non-global IP addresses.
+#[derive(Debug, Clone, Default)]
+pub struct GlobalIpOnly<T> {
+    inner: T,
+}
+
+mod ipv4_global {
```

Comment:
Do I understand correctly that these implementations are copied from `std`? If so, mind adding a comment above giving credit to the `std` authors and adding a link to the unstable tracking issue?

Reply:
@mxinden Some of the functions present in it are linked to `std` but some are not. Indeed, the `is_global` function is unstable and it does not seem to contain an exhaustive list of addresses to eliminate (all Special-Purpose Addresses) to know if our address is global. 

Thus, I referred to the exhaustive list of addresses, shared for example by the [Python implementation](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_global:~:text=(for%20IPv6).-,is_global%C2%B6,-True%20if%20the) of the function which it seems stable:
- `ipv4` list: https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml#note2
- `ipv6` list: https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml

To conclude, it is a mix between `std` functions and others that I have created. Do you want me to add a comment above all the ones that come from `std`?

Reply:
@mxinden And for specific address check functions that are declared as stable by `std`, I simply called the function directly from `std` without copying it back to the modules.

Reply:
Thanks for the quick reply!

Can you expand on what the Rust `std` implementation is missing? Do I understand correctly that we would never switch to the upstream Rust `std` `is_global` method, even when it is no longer unstable?

Reply:
@mxinden Of course, for example for the `ipv4` case, the `std` implementation is as follows:

```rust 
pub const fn is_global(&self) -> bool {
    !(self.octets()[0] == 0 // "This network"
        || self.is_private()
        || self.is_shared()
        || self.is_loopback()
        || self.is_link_local()
        // addresses reserved for future protocols (`192.0.0.0/24`)
        ||(self.octets()[0] == 192 && self.octets()[1] == 0 && self.octets()[2] == 0)
        || self.is_documentation()
        || self.is_benchmarking()
        || self.is_reserved()
        || self.is_broadcast())
}
```

**Taken into account by `std`:**
- `is_private()`: 10.0.0.0/8
- `is_shared()`: 100.64.0.0/10
- `is_loopback()`: 127.0.0.0/8
- `is_link_local()`: 169.254.0.0/16
- `(self.octets()[0] == 192 && self.octets()[1] == 0 && self.octets()[2] == 0)`: 192.0.0.0/24
- `is_documentation()`: 192.0.2.0/24, 198.51.100.0/24, 203.0.113.0/24
- `is_benchmarking()`: 198.18.0.0/15
- `is_reserved()`: 240.0.0.0/4
- `is_broadcast()`: 255.255.255.255

In the following [link](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml#note2), you will find the complete list of addresses to take into account according to the official specification, there are quite a few to add compared to the `std` implementation (and the number is even more important for the `ipv6` case).

Reply:
I understand the goal of being compliant with the official IANA specification. That said, I would assume that what works for the Rust `std` should as well work for rust-libp2p.

Can you help me understand in which way the Rust `std` implementation is lacking behind with IANA, where it is reasonable to expect that rust-libp2p deployments will run into the discrepancy?

The reason why I am being nit-picky here is that we will have to maintain this logic. While unlikely to change, it is a lot easier to maintain an exact copy of the Rust `std` implementation compared to a custom implementation.

One example, where I think it is fine to fall behind the IANA specification as the Rust `std` implementation does is the AS112 addresses (https://www.rfc-editor.org/rfc/rfc7535.html). How likely is it that a rust-libp2p deployment would actually see this address in the wild? And even if so, is it worth the additional code complexity and maintenance overhead introduced by diverging from the Rust `std` implementation?

Reply:
Yes indeed, I agree that one should add as little code complexity as possible in order to reduce the maintenance problems one might have. 

In fact, this part has been declared unstable by `std` since already 2015. So I thought it was possible that they forgot or simply didn't have time to implement certain addresses (maybe also the reason why they declare this unstable part because in reality there are not too many reasons, the functions are independent and extremely simple).

I looked on the IANA document how often they add specific new addresses and saw that the last addition was in 2017 (for `ipv4` for example). So I figured it doesn't happen that often and one could maintain a stable version for quite a while without patching.

But at the end, we can see the problem in two ways:
1. We want to be really exhaustive and we keep the list that I established in the code, telling ourselves that we will not touch it anytime soon (or maybe even by then we will have a stable version of `is_global` directly in `std`).
2. We want to stick exactly to `std` for all the constraints mentioned above and in this case I can copy and paste the `std` functions, there is no problem.

Reply:
I'd vote for (2) and perhaps we can lobby over at std to stabilize this function. Long-term I'd really like to get rid of that code.

This gives me an idea, to externalize this into a crate and I found: https://docs.rs/ip_rfc

Do we want to use that instead until the code in stdlib hits stable?

Reply:
@thomaseizinger Of course, I agree with you, it is absolutely necessary to get rid of all these lines of code: it's too many cases to maintain, it must be done by an external library.

No worries for option 2, if both of you @thomaseizinger @mxinden confirm this option, I can take care of duplicating the standard library. 

The `ip_rfc` library seems to provide the exact copy of the `std` library, so that might also be an option to consider (I looked at the functions, it seems to be exactly the same as `std`).

Reply:
I think depending on `ip_rfc` might be a nice mid-term solution. What do you think @mxinden?

Reply:
Are we aware of any case where the current `std` implementation falls short? "Short" not as in missing cases documented in an RFC, but as in missing cases that we would ever encounter in the wild libp2p world? If not, for the sake of not introducing another moving part (i.e. a new dependency), I am still advocating for simply copying the `std` implementation.

That said, looking at `ip_rfc`, this is a viable option to me as well, though I still prefer vanilla (2). @tcoratger please proceed either with (2) or use `ip_rfc`, whichever option you prefer. Thanks for your help!

Reply:
@thomaseizinger @mxinden No problem, following your recommendations:
- I implemented the exact copy of `std` for both `ipv4` and `ipv6` standards
- I let the functions declared as stable by `std` directly linked to `std`
- I implemented in two modules the remaining functions that are declared unstable by `std`
- I pushed a comment on the top of the modules to give credits to `std` team and a link to the unstable issue.